### PR TITLE
Copy-edit Chapter 2 up to the end of 2.3

### PR DESCRIPTION
**Pull request by tbrunz**
Hi @guillep , I realized that the Marshalling chapter made no mention of the issues regarding the bit size of the Pharo image (and the host OS bit size), and I think it should...

So I ran the examples in both a 32-bit and 64-bit P8 image (on Ubuntu 18.04 64-bit), and --as you would expect-- got very different results for the two regarding the cases where type conversion has issues.  I think this is a good thing, because it really emphasizes the point being made: You'll get strange, incorrect, and unpredictable results if you don't pay close attention to getting the type declarations correct.

But it also brings up another issue: You still need to be aware of the limitations of the bit size of your Pharo image & host OS, and how they will interact with the "personality" of C types when marshalling occurs.  It can be counter-intuitive sometimes, too.

I see you've started a new branch... I hope this doesn't cause too many conflicts.  (This is just Chapter 2; I'm about to start copy-editing Section 2.4.)

### CODE REVIEW

**Comment by guillep** [comment]
> Hi @guillep , I realized that the Marshalling chapter made no mention of the issues regarding the bit size of the Pharo image (and the host OS bit size), and I think it should...
> 
> So I ran the examples in both a 32-bit and 64-bit P8 image (on Ubuntu 18.04 64-bit), and --as you would expect-- got very different results for the two regarding the cases where type conversion has issues. I think this is a good thing, because it really emphasizes the point being made: You'll get strange, incorrect, and unpredictable results if you don't pay close attention to getting the type declarations correct.

Yes, you're right, actually the example of the 31 bit small integers is only valid for 32 bits, for 64 bits small integers are 61 bits long IIRC, we could check... We could add a not on that, but I don't think it's so interesting for the reader as long as uffi does the marshalling correctly.

I plan to add at the end of the section a table of basic C types supported by uffi + their sizes, and what types get marshalled to them and how. For example, booleans can be marshalled to ints 0 and 1, nil can be marshalled to null in a pointer type (I don't remember if nil = 0 if type = int :P)...


> But it also brings up another issue: You still need to be aware of the limitations of the bit size of your Pharo image & host OS, and how they will interact with the "personality" of C types when marshalling occurs. It can be counter-intuitive sometimes, too.

Can you give an example? To me in general you should really care about the C-side types but on the smalltalk size it should be transparent: for example, small integers differ in size yes, but they get automatically transformed in large integers as soon as you need it, and it's completely transparent for the user regardless the platform and architecture.



> I see you've started a new branch... I hope this doesn't cause too many conflicts. (This is just Chapter 2; I'm about to start copy-editing Section 2.4.)

That's a mistake, It's a branch to make a PR only, I should have done it in my fork but baby started crying :)

File: Chapters/2-marshalling.pillar [hunk 1]
```diff
@@ -1,96 +1,123 @@
...
+[[[language=smalltalk
+FFITutorial abs: SmallInteger maxVal negated.
+=> 1073741823
+]]]
+
+- ==SmallInteger==s in 64-bit Pharo images ''can'' overflow the size of a C ==int== (still 32 bits on most 64-bit hosts), and so are coerced by truncating their value to fit, producing results similar to what a C program would produce in a similar situation.  Since the ==maxVal== of a 64-bit Pharo ==SmallInteger== is 60 \'1\' bits (plus a sign bit), it truncates to 32 \'1\' bits, which, to C, is the two\'s complement value -1.  Hence,
 
 [[[language=smalltalk
-FFITutorial abs: SmallInteger maxVal.  "Cast to all 1's"
+FFITutorial abs: SmallInteger maxVal negated.
```
Comment (guillep) [inline Chapters/2-marshalling.pillar]: Yes, nice example!

Now, this assumes the reader knows that ==int==s are 32bit sized :) Maybe we can add a clarification on this?

Comment (guillep) [reply Chapters/2-marshalling.pillar]: And also the fact that ==int== is signed, thus 1 bit is there for the sign :)

Comment (guillep) [reply Chapters/2-marshalling.pillar]: (which is necessary to understand the explanation on the return value in the next subsection ;))

**Comment by guillep** [comment]
I took the liberty of merging your branch with my latest changes.
I'll now open a new section on arguments by copy and by reference.

**Comment by tbrunz** [comment]
Just got up...  I think it may be easier to edit from here after generating a new PDF.  (Especially since there are now 2 sets of edits combined.)

**Comment by guillep** [comment]
ok!

File: Chapters/2-marshalling.pillar (hunk 1 above)
Comment (tbrunz) [reply Chapters/2-marshalling.pillar]: Yes, I've been trying to add notes to remind the reader that C int's are 32 bit regardless of the Pharo image bit size.  And I mentioned that "all 1s" is two's complement of "-1". 

One thing we might mention is that Pharo keeps a separate sign bit, too.  I found this out while experimenting: `maxVal` and `maxVal negated` have the _same_ number of value bits, which is not like C 2's complement values.  The reader probably should know this, too.

**Comment by tbrunz** [comment]
> > But it also brings up another issue: You still need to be aware of the limitations of the bit size of your Pharo image & host OS, and how they will interact with the "personality" of C types when marshalling occurs. It can be counter-intuitive sometimes, too.

> 
> Can you give an example? To me in general you should really care about the C-side types but on the smalltalk size it should be transparent: for example, small integers differ in size yes, but they get automatically transformed in large integers as soon as you need it, and it's completely transparent for the user regardless the platform and architecture.

Example: `SmallInteger * -10` will be implicitly converted within Pharo to a `LargeNegativeInteger`, but then instead of being coerced by uFFI to a C int, it blows up: You get a "Cannot coerce arguments" error instead of a truncated integer.  (And I think the behavior is different in 32-bit P8, but my test image is at work, not home. :^)

I expanded the examples in this section with 32-bit and 64-bit "versions", which shows this (and a couple of other cases, including a convenient "float nan" result).  I think you'll like the results...

**Comment by guillep** [comment]
> Example: `SmallInteger * -10` will be implicitly converted within Pharo to a `LargeNegativeInteger`, but then instead of being coerced by uFFI to a C int, it blows up: You get a "Cannot coerce arguments" error instead of a truncated integer. (And I think the behavior is different in 32-bit P8, but my test image is at work, not home. :^)
> 
> I expanded the examples in this section with 32-bit and 64-bit "versions", which shows this (and a couple of other cases, including a convenient "float nan" result). I think you'll like the results...

Yup, I saw that later while reading your changes :)

**Comment by tbrunz** [comment]
I see better proofing the PDFs, which is why I pressed Stef for a way to download the "test build" PDFs that Travis makes.  I can see at the end of its log that it runs the "upload" bash script, but I don't know where it "goes", nor how to access it / download it, to proofread (before an actual release).  Is this possible??

**Comment by guillep** [comment]
Well, the pdf is uploaded to the releases of the repository. Check in the main page, next to the number of commits and branches, there are the releases. However, only the main branch is uploaded if I'm not mistaken, so changes should be integrated before being in the pdf.

image

Another way to do it is to locally install pillar + latex and then run `pillar build pdf unified-ffi.pillar`.
Either way, I can try to merge fast enough so we have the latest pdf soon enough to get feedback.

**Comment by guillep** [comment]
Check in https://github.com/SquareBracketAssociates/Booklet-uFFI/releases the **assets** within continuous build.

**Comment by tbrunz** [comment]
Yes, but those **assets** only show up when a release is actually performed.  Until the release is done, the PDF isn't visible.  (And of course I can't make a release!)

I am meaning to install Pillar, LaTex, etc. and build them myself; I just haven't gotten to it yet.  :^)

Anyway, I see the release you just made, so I'll download it and go through it.  Thanks!

**Pull request merged**